packages/caravan-health/src/wallet.ts

```diff
@@ -0,0 +1,131 @@
+    - 0.5 for UTXO set length >= 15 and <= 24
+    - 0.75 for UTXO set length >= 5 and <= 14
+    - 1 for UTXO set length < 5
+  */
+  utxoMassFactor(utxos: AddressUtxos): number {
+    let utxoSetLength = 0;
+    for (const address in utxos) {
+      const addressUtxos = utxos[address];
+      utxoSetLength += addressUtxos.length;
+    }
+    let utxoMassFactor: number;
+    if (utxoSetLength >= 50) {
+      utxoMassFactor = 0;
+    } else if (utxoSetLength >= 25 && utxoSetLength <= 49) {
+      utxoMassFactor = 0.25;
+    } else if (utxoSetLength >= 15 && utxoSetLength <= 24) {
+      utxoMassFactor = 0.5;
+    } else if (utxoSetLength >= 5 && utxoSetLength <= 14) {
+      utxoMassFactor = 0.75;
+    } else {
+      utxoMassFactor = 1;
+    }
+    return utxoMassFactor;
+  }
+
+  /*
+    Utility function that helps to obtain the fee rate of the transaction
+  */
+  getFeeRateForTransaction(transaction: Transaction): number {
+    let fees: number = transaction.fee;
+    let weight: number = transaction.weight;
+    return fees / weight;
+  }
+
+  /*
+    Utility function that helps to obtain the percentile of the fees paid by user in tx block
+  */
+  getFeeRatePercentileScore(
+    timestamp: number,
+    feeRate: number,
+    feeRatePercentileHistory: FeeRatePercentile[],
+  ): number {
+    let percentile: number = this.getClosestPercentile(
+      timestamp,
+      feeRate,
+      feeRatePercentileHistory,
+    );
+    return 1 - percentile / 100;
+  }
+
+  /*
+    Utility function that helps to obtain the closest percentile of the fees paid by user in tx block
+  */
+  getClosestPercentile(
+    timestamp: number,
+    feeRate: number,
+    feeRatePercentileHistory: FeeRatePercentile[],
+  ): number {
+    // Find the closest entry by timestamp
+    let closestBlock: FeeRatePercentile | null = null;
+    let closestDifference: number = Infinity;
+
+    for (const block of feeRatePercentileHistory) {
+      const difference = Math.abs(block.timestamp - timestamp);
+      if (difference <= closestDifference) {
+        closestDifference = difference;
+        closestBlock = block;
+      }
+    }
+    if (!closestBlock) {
+      throw new Error("No fee rate data found");
+    }
+    // Find the closest fee rate percentile
+    switch (true) {
+      case feeRate <= closestBlock.avgFee_0:
+        return 0;
+      case feeRate <= closestBlock.avgFee_10:
+        return 10;
+      case feeRate <= closestBlock.avgFee_25:
+        return 25;
+      case feeRate <= closestBlock.avgFee_50:
+        return 50;
+      case feeRate <= closestBlock.avgFee_75:
+        return 75;
+      case feeRate <= closestBlock.avgFee_90:
+        return 90;
+      case feeRate <= closestBlock.avgFee_100:
+        return 100;
+      default:
+        throw new Error("Invalid fee rate");
+    }
+  }
+
+  constructAddressUsageMap(transactions: Transaction[]): AddressUsageMap {
+    let addressUsageMap: AddressUsageMap = {};
```

Comment:
if you wanted to cache these you'd want to save this on the class. Then it could be reused. to get into the other hardest thing in computer science, you might also want cache invalidation in case transactions were added, but that's probably a premature optimization.

Reply:
Also similarly to the comment below, if you instantiated this class with a list of transactions (which I think makes sense since it has "wallet" in the name) then you can just call this on `this.transactions`